ipfsconn/ipfshttp/ipfshttp.go

```diff
@@ -450,6 +459,97 @@ func checkResponse(path string, code int, body []byte) error {
 	return fmt.Errorf("IPFS-post '%s' unsuccessful: %d: %s", path, code, body)
 }
 
+// postCtxStreaming makes a POST request against
+// the ipfs daemon, reads the body of the response line by line
+// at evey millisecond. It cancels the context if timeout duration
+// has passed since last new data on the buffer.
+func (ipfs *Connector) postCtxStreaming(ctx context.Context, path string, timeout time.Duration) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	timer := time.NewTimer(timeout)
+	defer timer.Stop()
+
+	elapsed := time.Now()
+
+	go func() {
+		for {
+			select {
+			case <-timer.C:
+				{
+					if time.Since(elapsed) >= timeout {
+						cancel()
+						break
+					}
+					timer.Reset(timeout - time.Since(elapsed))
+				}
+			}
+		}
+	}()
+
+	res, err := ipfs.doPostCtx(ctx, ipfs.client, ipfs.apiURL(), path, "", nil)
+	if err != nil {
+		return err
+	}
+	defer res.Body.Close()
+	reader := bufio.NewReader(res.Body)
+
+	if strings.HasPrefix(path, "pin/add") {
+		var pins ipfsPinsResp
+		var progress int
+		for {
+			line, err := reader.ReadBytes('\n')
+			if err != nil {
+				if err == io.EOF {
+					err = json.Unmarshal(line, &pins)
+					break
+				}
+				return err
+			}
+
+			err = json.Unmarshal(line, &pins)
+			if err != nil {
+				return err
+			}
+			if pins.Progress > progress {
+				progress = pins.Progress
+				elapsed = time.Now()
```

Comment:
this is racy, chances that it will panic

Reply:
https://github.com/ipfs/ipfs-cluster/pull/738/commits/58c3a98bd444ef85774296b2504863186b726600

Reply:
Not fixed, time.Reset could also lead to racy behavior. Will add changes

Reply:
This should take care of racy behaviour on using timer.Reset https://github.com/ipfs/ipfs-cluster/pull/738/commits/914b270c993bd373618fd39c5d6992e971ee6c16